### PR TITLE
Scoped WebAuthn: Refactor `SessionData`

### PR DESCRIPTION
**Pull request by Joerger**
Moves custom webauthn `SessionData` struct out of `webauthn.proto` to an internal implementation. This struct was previously only used to standardize json marshalling/unmarshalling for backend storage. Now it also includes Teleport extensions.

This was [requested](https://github.com/gravitational/teleport/pull/36588#discussion_r1450848226) as part of the work for [Scoped Webauthn Challenges](https://github.com/gravitational/teleport/pull/35185).

### CODE REVIEW

File: lib/auth/webauthn/session.go [hunk 1]
```diff
@@ -39,21 +39,40 @@ const scopeLogin = "login"
 // that use in-memory storage.
 const scopeSession = "registration"
 
-func sessionToPB(sd *wan.SessionData) (*wanpb.SessionData, error) {
+// SessionData is a clone of [protocol.SessionData], materialized here to keep a
+// stable JSON marshal/unmarshal representation and add extensions.
+type SessionData struct {
```
Comment (codingllama) [inline lib/auth/webauthn/session.go:44]: Move it to lib/auth/webauthntypes, so we avoid eventual cycles/importing the entirety of server-side webauthn just for this type?

File: lib/auth/webauthn/session.go [hunk 2]
```diff
@@ -39,21 +39,40 @@ const scopeLogin = "login"
...
+	// Raw Credential IDs of the credentials allowed for the ceremony.
+	AllowCredentials [][]byte `json:"allowCredentials,omitempty"`
+	// True if resident keys were required by the server / Relying Party.
+	ResidentKey bool `json:"residentKey,omitempty"`
+	// Requested user verification requirement, either "discouraged" or
+	// "required".
+	// An empty value is treated equivalently to "discouraged".
+	UserVerification string `json:"userVerification,omitempty"`
+	// Extensions are Teleport extensions that apply to this webauthn session.
+	Extensions *proto.MFAChallengeExtensions `json:"extensions,omitempty"`
```
Comment (codingllama) [inline lib/auth/webauthn/session.go:58]: `Extensions` already has [meaning in WebAuthn][1] and it something we could add in the future. Let's call this something more specific, `ChallengeExtensions` maybe?

[1]: https://pkg.go.dev/github.com/go-webauthn/webauthn/webauthn#SessionData

File: lib/auth/webauthn/session.go [hunk 3]
```diff
@@ -39,21 +39,40 @@ const scopeLogin = "login"
...
+}
+
+func sessionDataFromWebauthnProtocol(sd *wan.SessionData) (*SessionData, error) {
```
Comment (codingllama) [inline lib/auth/webauthn/session.go:61]: Suggestion: `sessionDataToProtocol` and `sessionDataFromProtocol` seem good enough to me.

File: lib/auth/webauthn/login_test.go [hunk 4]
```diff
@@ -412,22 +410,19 @@ func TestPasswordlessFlow_BeginAndFinish(t *testing.T) {
...
 				UserId:           nil,   // aka unset
 				AllowCredentials: nil,   // aka unset
 				ResidentKey:      false, // irrelevant for login
 				UserVerification: string(protocol.VerificationRequired),
 			}
-			if !proto.Equal(sd, wantSD) {
-				diff := cmp.Diff(wantSD, sd)
-				t.Fatalf("SessionData mismatch (-want +got):\n%s", diff)
-			}
+			require.Equal(t, wantSD, sd, "SessionData mismatch")
```
Comment (codingllama) [inline lib/auth/webauthn/login_test.go:425]: Maybe:

```suggestion
			if diff := cmp.Diff(wantSD, sd); diff != "" {
				t.Fatalf("SessionData mismatch (-want +got):\n%s", diff)
			}
```

Closer to the original, plus I like the diff-error better than the requires error... 😊

File: lib/auth/webauthn/session.go [hunk 5]
```diff
@@ -25,7 +25,7 @@ import (
 	wan "github.com/go-webauthn/webauthn/webauthn"
 	"github.com/gravitational/trace"
 
-	wanpb "github.com/gravitational/teleport/api/types/webauthn"
+	"github.com/gravitational/teleport/api/client/proto"
```
Comment (codingllama) [inline lib/auth/webauthn/session.go:28]: Maybe:

```suggestion
	authpb "github.com/gravitational/teleport/api/client/proto"
```

`proto` is a terrible package name, plus it confuses itself with the official proto imports.

**Review by rosstimothy — APPROVED** [review]

**Review by GavinFrazar — APPROVED** [review]

File: lib/auth/webauthntypes/webauthn.go [hunk 6]
```diff
@@ -376,3 +379,46 @@ type Credential protocol.Credential
...
+	// Raw Credential IDs of the credentials allowed for the ceremony.
+	AllowCredentials [][]byte `json:"allowCredentials,omitempty"`
+	// True if resident keys were required by the server / Relying Party.
+	ResidentKey bool `json:"residentKey,omitempty"`
+	// Requested user verification requirement, either "discouraged" or
+	// "required".
+	// An empty value is treated equivalently to "discouraged".
+	UserVerification string `json:"userVerification,omitempty"`
+	// ChallengeExtensions are Teleport extensions that apply to this webauthn session.
+	ChallengeExtensions mfav1.ChallengeExtensions `json:"challenge_extensions,omitempty"`
```
Comment (codingllama) [inline lib/auth/webauthntypes/webauthn.go:399]: Maybe, so we can differentiate set/unset:

```suggestion
	ChallengeExtensions *mfav1.ChallengeExtensions `json:"challenge_extensions,omitempty"`
```

It doesn't look like this is needed now, but the pointer is the more "usual" design.

File: lib/auth/webauthntypes/webauthn.go [hunk 7]
```diff
@@ -376,3 +379,46 @@ type Credential protocol.Credential
...
+}
+
+func SessionDataFromProtocol(sd *webauthn.SessionData) (*SessionData, error) {
```
Comment (codingllama) [inline lib/auth/webauthntypes/webauthn.go:402]: nit: godoc.

**Review by codingllama — APPROVED** [review]
Please address the comments, otherwise LGTM.

**Comment by codingllama** [comment]
It also looks like you have a few test failures.

**Comment by Joerger** [comment]
@r0mant @zmb3 Can i get a flaky test skip? See https://github.com/gravitational/teleport/issues/36832

File: lib/auth/webauthntypes/webauthn.go [hunk 8]
```diff
@@ -376,3 +379,50 @@ type Credential protocol.Credential
 type AuthenticationExtensionsClientOutputs struct {
 	AppID bool `json:"appid,omitempty"`
 }
+
+// SessionData is a clone of [webauthn.SessionData], materialized here to keep a
+// stable JSON marshal/unmarshal representation and add extensions.
```
Comment (ibeckermayer) [inline lib/auth/webauthntypes/webauthn.go:384]: This seems to have diverged from the version of the `webauthn.SessionData` we're using (it's missing `Expires` and `Extensions`)

https://pkg.go.dev/github.com/go-webauthn/webauthn@v0.10.0/webauthn#SessionData

(Unclear to me whether that's a bug in the code or the comment)

Comment (Joerger) [reply lib/auth/webauthntypes/webauthn.go:384]: There's a TODO from @codingllama to add extensions. I'll update and move it to mention expires as well. IIUC we are not using either of these fields currently so we don't have a reason to record them as of now.

Comment (codingllama) [reply lib/auth/webauthntypes/webauthn.go:384]: It's alright, we don't need every feature from upstream. Other messages probably have small divergences too.

**Comment by Joerger** [comment]
@codingllama I am still getting flaky test issues with `TestIdentityService_UpsertGlobalWebauthnSessionData_maxLimit`, though I can no longer reproduce it locally. Any ideas?

**Comment by codingllama** [comment]
It looks like that when you run multiple tests they interact in a bad way. Let me see what I can do.

**Comment by Joerger** [comment]
@r0mant @zmb3 Can I please get a flaky test skip to unblock this?

**Comment by codingllama** [comment]
Flaky fix at https://github.com/gravitational/teleport/pull/36945.

**Pull request merged**